
---

# Fundamentos de Procesamiento Textual  
## Limpieza, estructuración y representación de un corpus

---

## 🌐 ¿Qué es el Procesamiento de Lenguaje Natural (NLP)?

El **Procesamiento de Lenguaje Natural (Natural Language Processing, NLP)** es un área interdisciplinaria que combina lingüística, estadística, aprendizaje automático e inteligencia artificial para permitir que las computadoras analicen, interpreten y generen lenguaje humano.

Tradicionalmente, el NLP se basaba en representaciones estadísticas del texto, tales como frecuencias de palabras, matrices documento-término y modelos probabilísticos. En los últimos años, el desarrollo de los **Large Language Models (LLM)** y de las arquitecturas **Transformer** ha ampliado significativamente las capacidades del campo, permitiendo comprender relaciones semánticas complejas, generar texto y realizar tareas de razonamiento sobre lenguaje.

El objetivo central del NLP sigue siendo el mismo: **transformar información textual no estructurada en conocimiento útil y analizable**. Sin embargo, hoy disponemos de múltiples enfoques para lograrlo, cada uno con ventajas y limitaciones.

---

## 🛠️ Aplicaciones del NLP que veremos en este curso

Durante este curso estudiaremos tanto enfoques clásicos como modernos de NLP utilizando **R**, **Python** y modelos basados en **Transformers**.

### NLP Tradicional (R / Quanteda)

- Limpieza y transformación de textos.
- Tokenización, normalización y manejo de stopwords.
- Construcción de matrices documento-término (DFM).
- Estadística descriptiva textual.
- Análisis de keyness y dispersión léxica.
- Modelamiento de tópicos mediante LDA.
- Similitud textual basada en TF-IDF.

### NLP Moderno (Python / Transformers)

- Representaciones vectoriales mediante embeddings.
- Clasificación de textos utilizando modelos Transformer.
- Análisis de sentimiento con modelos especializados.
- Similitud semántica mediante *sentence-transformers*.
- Descubrimiento de tópicos con BERTopic.
- Uso de LLM para interpretación, codificación y análisis asistido de textos.

---

## ⚖️ ¿Por qué aprender ambos enfoques?

Una idea central de este curso es que **no existe una herramienta universalmente superior**.

Los métodos tradicionales suelen ofrecer:

- Mayor interpretabilidad.
- Menor costo computacional.
- Alta reproducibilidad.
- Excelente desempeño en tareas descriptivas.

Los enfoques basados en Transformers y LLM suelen ofrecer:

- Mejor comprensión contextual.
- Mayor capacidad semántica.
- Flexibilidad para tareas complejas.
- Desempeño superior en clasificación y análisis contextual.

Aprenderemos cuándo utilizar cada enfoque y cómo comparar sus resultados de manera crítica y rigurosa.

> 📌 El NLP se utiliza actualmente en economía, finanzas, marketing, ciencias sociales, análisis político, medios de comunicación, salud, derecho y numerosos otros campos donde el texto constituye una fuente relevante de información.

## 🤖 NLP Tradicional vs Modelos de Lenguaje de Gran Escala (LLMs)

### 🧠 ¿Qué es el NLP?

El **Procesamiento de Lenguaje Natural (NLP)** es un campo de la inteligencia artificial que permite a las computadoras analizar, comprender y generar lenguaje humano. Combina herramientas estadísticas, lingüísticas y computacionales para transformar texto no estructurado en información útil y analizable.

Entre sus aplicaciones más comunes se encuentran:

- Limpieza y transformación de texto.
- Análisis de frecuencia y coocurrencia.
- Construcción de matrices documento-término.
- Modelamiento de tópicos (LDA).
- Análisis de sentimiento.
- Medición de similitud entre documentos.
- Clasificación automática de textos.

---

### 🚀 ¿Qué son los LLMs?

Los **Modelos de Lenguaje de Gran Escala (Large Language Models, LLMs)** son modelos de inteligencia artificial basados en la arquitectura **Transformer**, entrenados sobre enormes colecciones de texto.

Gracias a este entrenamiento, pueden capturar relaciones semánticas complejas y realizar múltiples tareas relacionadas con el lenguaje, tales como:

- Responder preguntas.
- Generar texto.
- Resumir documentos.
- Traducir entre idiomas.
- Clasificar textos.
- Extraer información relevante.

Modelos como GPT, Claude, Gemini, Llama o Mistral representan algunos de los ejemplos más conocidos de esta nueva generación de herramientas.

---

### ✅ ¿Para qué es mejor cada enfoque?

#### NLP Tradicional

Es especialmente útil cuando:

- Se requiere transparencia e interpretabilidad.
- Se desea controlar cada etapa del procesamiento.
- Se trabaja con corpus grandes y bien estructurados.
- Se busca máxima reproducibilidad metodológica.
- Se dispone de recursos computacionales limitados.

#### LLMs

Son especialmente útiles cuando:

- Se requiere comprensión contextual profunda.
- Los textos contienen ambigüedad, jerga o lenguaje complejo.
- Se necesitan tareas de generación de texto o resumen automático.
- No se dispone de datos etiquetados para entrenamiento.
- Se busca flexibilidad para resolver múltiples tareas con un mismo modelo.

---

### 📊 Comparación de Ventajas y Desventajas

| Característica | NLP Tradicional | LLMs |
|----------------|----------------|-------|
| Interpretabilidad | Alta | Baja |
| Requerimientos computacionales | Bajos | Altos |
| Reproducibilidad | Alta | Moderada |
| Escalabilidad | Muy alta | Alta, pero costosa |
| Comprensión contextual | Limitada | Muy alta |
| Necesidad de datos etiquetados | Frecuentemente necesaria | Puede operar con *zero-shot* o *few-shot* |
| Control metodológico | Alto | Menor |
| Costo de implementación | Bajo | Variable |

---

### 🎯 Conclusión

Los enfoques tradicionales y los LLMs deben entenderse como herramientas complementarias más que como alternativas excluyentes.

Los métodos clásicos continúan siendo fundamentales para tareas descriptivas, análisis reproducibles y estudios académicos rigurosos. Los LLMs, por su parte, ofrecen capacidades superiores para comprender contexto, interpretar significado y automatizar tareas complejas.

> 📌 A lo largo del curso aprenderemos ambos enfoques y discutiremos cuándo resulta más apropiado utilizar cada uno.

---
---

# I. Cargando **"quanteda"** y nuestro primer ***corpus***

## 📦 ¿Qué es `quanteda`?

`quanteda` (Quantitative Analysis of Textual Data) es un **paquete de R** diseñado para facilitar el análisis cuantitativo de textos. Es una herramienta poderosa, eficiente y flexible que permite transformar texto no estructurado en objetos estructurados para su análisis estadístico.

---

### 🧰 Funcionalidades principales

- **Creación y manipulación de corpus**  
  Permite importar, visualizar y gestionar grandes colecciones de textos.

- **Tokenización y limpieza de texto**  
  Remueve puntuación, stopwords, números, símbolos y permite hacer stemming y lematización.

- **Construcción de la Document-Feature Matrix (DFM)**  
  Convierte el texto en una matriz estructurada donde las filas son documentos y las columnas son términos.

- **Análisis de frecuencia y coocurrencia**  
  Mide cuán frecuentemente ocurren palabras o aparecen juntas.

- **Visualizaciones**  
  Incluye funciones para generar nubes de palabras, gráficos de dispersión léxica, keyness, etc.

- **Modelado de tópicos y análisis avanzado**  
  Se puede integrar fácilmente con otros paquetes como `topicmodels`, `tidytext`, o `stm`.

---

### 💡 ¿Por qué usar `quanteda`?

- Optimizado para **grandes volúmenes de texto**
- Basado en estructuras de datos eficientes (`dfm`)
- Compatible con el **ecosistema tidyverse**
- Ideal para investigadores en **ciencias sociales, lingüística computacional, economía y comunicación**

---

### 🧪 Ejemplos de aplicaciones (en este curso)

- Análisis de sentimiento en reseñas de películas
- Identificación de tópicos con LDA
- Detección de palabras clave por grupo (keyness)
- Visualización de patrones léxicos y semánticos
- Comparación de documentos por similitud

> En resumen, `quanteda` es una de las herramientas más completas en R para realizar NLP clásico de forma transparente, reproducible y eficiente.


In [1]:
install.packages("quanteda")
library(quanteda)
library(stringi)

# En sistemas Mac, fijamos el locale para evitar problemas de codificación con tildes y ñ
Sys.setlocale("LC_CTYPE", "en_US.UTF-8")



The downloaded binary packages are in
	/var/folders/8q/0f19vqjs5hn942n9xdsj2lc80000gn/T//RtmpBke1EA/downloaded_packages


Warning message:
“package ‘quanteda’ was built under R version 4.5.2”
Package version: 4.4
Unicode version: 14.0
ICU version: 71.1

Parallel computing: disabled

See https://quanteda.io for tutorials and examples.



[1] "en_US.UTF-8"

## 📚 ¿Qué es un *Corpus*?

En lingüística y procesamiento del lenguaje natural (NLP), un **corpus** es una colección organizada de documentos textuales que constituye la unidad básica de análisis. Los documentos pueden provenir de múltiples fuentes, tales como noticias, artículos académicos, redes sociales, informes financieros, documentos legales o transcripciones de entrevistas.

Los corpus permiten estudiar patrones lingüísticos, extraer información y entrenar modelos capaces de analizar o generar lenguaje humano.

---

### 🧩 Características de un corpus

- Está compuesto por uno o más documentos.
- Puede contener textos escritos o transcripciones de lenguaje hablado.
- Permite analizar el lenguaje en contextos reales.
- Puede estar acompañado de **metadatos** (*document variables* o *docvars*), tales como:
  - Fecha de publicación.
  - Autor o fuente.
  - Categoría temática.
  - País o ubicación.
  - Etiquetas de sentimiento o clasificación.

---

### 🛠️ ¿Para qué sirve un corpus?

Los corpus constituyen la materia prima de prácticamente todas las aplicaciones modernas de NLP:

- Análisis de frecuencias y vocabulario.
- Identificación de patrones lingüísticos.
- Análisis de sentimiento.
- Modelamiento de tópicos.
- Medición de similitud entre documentos.
- Entrenamiento y evaluación de modelos de clasificación.
- Construcción de embeddings y aplicaciones basadas en LLMs.

Sin un corpus adecuadamente construido y documentado, incluso los modelos más sofisticados producirán resultados poco confiables.

---

### 📦 Corpus en Quanteda

En `quanteda`, un corpus es un objeto estructurado que almacena tanto los textos como sus metadatos asociados. Esta estructura permite construir flujos de trabajo reproducibles para transformar texto no estructurado en datos analizables.

> 📌 En este curso, todos los análisis comienzan con la construcción de un corpus bien definido y documentado.

---

## 📂 Corpus: `data_corpus_inaugural`

El corpus `data_corpus_inaugural` contiene los discursos inaugurales pronunciados por los presidentes de Estados Unidos al asumir el cargo, lo cual ocurre cada cuatro años el 21 de enero. Estos discursos son documentos clave que suelen reflejar las intenciones, prioridades y visión del mandatario para el país.

---

### 🧪 ¿Por qué es útil este corpus?

- Ofrece una **colección histórica** de documentos políticos, ideal para estudiar la evolución del lenguaje político a lo largo del tiempo.
- Permite analizar cómo cambian los **temas**, **emociones** y **estilos retóricos** a través de las décadas.
- Es una fuente muy utilizada para:
  - **Análisis de sentimiento**
  - **Modelamiento de tópicos**
  - **Estudios de frecuencia léxica**
  - **Extracción de palabras clave (keyness)**
  - **Medición de similitud semántica entre presidentes o períodos**

---

> Este corpus está integrado directamente en el paquete `quanteda` y es una excelente base para introducir conceptos fundamentales del NLP aplicado a textos políticos.


In [2]:
print(data_corpus_inaugural)

Corpus consisting of 60 documents and 4 docvars.
1789-Washington :
"Fellow-Citizens of the Senate and of the House of Representa..."

1793-Washington :
"Fellow citizens, I am again called upon by the voice of my c..."

1797-Adams :
"When it was first perceived, in early times, that no middle ..."

1801-Jefferson :
"Friends and Fellow Citizens: Called upon to undertake the du..."

1805-Jefferson :
"Proceeding, fellow citizens, to that qualification which the..."

1809-Madison :
"Unwilling to depart from examples of the most revered author..."

[ reached max_ndoc ... 54 more documents ]


---

## print(corpus), docvars(corpus), summary(corpus), docnames(corpus)

**docvars:** Las "docvars" (variables de documento) son metadatos asociados a cada documento en un corpus. Representan información adicional o atributos que pueden estar relacionados con los textos contenidos en el corpus.

### ¿Qué son metadatos en el contexto de NLP?


Los **metadatos** son *datos sobre los datos lingüísticos*. Es decir, no forman parte del contenido textual principal, sino que proporcionan **información adicional que describe, estructura o contextualiza ese texto**.

---

### 📌 Tipos de metadatos en NLP

#### 1. **Metadatos estructurales**
- Organizan el texto o documento.
- Ejemplos:
  - Título, subtítulos, capítulos
  - Párrafos o secciones (`<body>`, `<abstract>`)
  - Formato del documento

----

#### 2. **Metadatos administrativos**
- Información técnica o bibliográfica.
- Ejemplos:
  - Autor, fecha de publicación
  - Fuente o URL
  - Idioma del texto
  - Tipo de archivo (PDF, HTML, TXT)

----

#### 3. **Metadatos lingüísticos**
- Se derivan mediante técnicas de NLP.
- Ejemplos:
  - Tokens y lemas (`"hablando"` → `"hablar"`)
  - Etiquetas gramaticales (POS - *Part of Speech* tags)
  - Entidades nombradas (NER - *Named Entity Recogbnition*) (personas, ubicaciones, organizaciones, fechas, monedas, etc..)
  - Análisis de sentimientos
  - Dependencias sintácticas

----

### 🧠 ¿Qué son los POS tags en NLP?

Los **POS tags** (Part-of-Speech tags) son **etiquetas de categorías gramaticales** que se asignan a cada palabra de un texto, indicando su función sintáctica, como sustantivo, verbo, adjetivo, etc.


---

### 📚 Ejemplo

Frase: *"Los estudiantes corren rápido."*

| Palabra       | POS Tag  | Categoría         |
|---------------|----------|-------------------|
| Los           | `DET`    | Determinante      |
| estudiantes   | `NOUN`   | Sustantivo        |
| corren        | `VERB`   | Verbo             |
| rápido        | `ADV`    | Adverbio          |
| .             | `PUNCT`  | Puntuación        |

---

### 🔤 Etiquetas POS comunes (Universal POS tags)

| Etiqueta | Categoría gramatical   | Ejemplo       |
|----------|------------------------|---------------|
| `NOUN`   | Sustantivo             | perro         |
| `VERB`   | Verbo                  | correr        |
| `ADJ`    | Adjetivo               | feliz         |
| `ADV`    | Adverbio               | rápido        |
| `PRON`   | Pronombre              | él, ella      |
| `DET`    | Determinante           | el, una       |
| `ADP`    | Preposición            | en, con       |
| `CONJ`   | Conjunción             | y, pero       |
| `NUM`    | Número                 | tres          |
| `PUNCT`  | Puntuación             | . , ;         |

---

### 🛠️ ¿Para qué sirven?

- Análisis sintáctico y gramatical
- Traducción automática
- Extracción de información
- Desambiguación de palabras
- Apoyo a tareas como NER y análisis de sentimiento

---

> ✨ **Conclusión:** El etiquetado POS permite comprender la estructura del lenguaje, y es fundamental para muchas tareas de procesamiento del lenguaje natural.

----

#### 4. **Metadatos contextuales o sociales**
- Situación en que se produce el texto.
- Ejemplos:
  - Usuario que escribió el texto
  - Fecha y lugar de publicación
  - Interacciones (likes, retweets)
  - Medio de comunicación (Twitter, blog, noticia)

---

### 📊 Ejemplo práctico

Supongamos que tienes este tweet:

> *“No puedo creer lo que pasó hoy en Santiago. ¡Qué rabia!”*

Los **metadatos** asociados podrían ser:

| Tipo                 | Ejemplo                                      |
|----------------------|-----------------------------------------------|
| Fecha                | 2025-07-22                                    |
| Usuario              | @juanito_chile                                |
| Localización         | Santiago, Chile                               |
| Sentimiento          | Negativo                                      |
| Entidades nombradas  | Santiago (Lugar)                              |
| POS Tags             | `No/ADV puedo/V creer/V lo/DET que/PR pasó/V`... |

---

### 🧠 ¿Para qué sirven los metadatos en NLP?

- **Filtrar o clasificar textos** (por autor, tema, idioma)
- **Entrenar modelos más precisos** (p. ej. ajustar por país o tono)
- **Interpretar resultados** (por ejemplo, evolución del sentimiento)
- **Recuperación de información avanzada** (búsqueda semántica, clustering)

---






> 🔍 **Conclusión:** Los metadatos enriquecen el análisis lingüístico y permiten usar el texto en tareas complejas como clasificación, extracción de información o análisis de discurso. En este sentido, las docvars son útiles para almacenar información contextual sobre los documentos, como el autor, el género, la fecha, la fuente o cualquier otro atributo relevante -*medadatp* - que se quiera asociar a los textos. Estas variables pueden ser de diferentes tipos, como cadenas de texto, fechas, números o factores categóricos.



In [3]:
head(docvars(data_corpus_inaugural))

,Year,President,FirstName,Party
,<int>,<chr>,<chr>,<fct>
1,1789,Washington,George,none
2,1793,Washington,George,none
3,1797,Adams,John,Federalist
4,1801,Jefferson,Thomas,Democratic-Republican
5,1805,Jefferson,Thomas,Democratic-Republican
6,1809,Madison,James,Democratic-Republican


**summary(corpus)**

Otra forma de obtener una visión general condensada de un corpus es utilizar la función **summary()**. **summary()** es una función base de R (es decir, una función incorporada en R) y proporciona información resumida básica sobre el objeto en cuestión.

In [4]:
head(summary(data_corpus_inaugural))
tail(summary(data_corpus_inaugural))

,Text,Types,Tokens,Sentences,Year,President,FirstName,Party
,<chr>,<int>,<int>,<int>,<int>,<chr>,<chr>,<fct>
1,1789-Washington,625,1537,24,1789,Washington,George,none
2,1793-Washington,96,147,5,1793,Washington,George,none
3,1797-Adams,826,2577,37,1797,Adams,John,Federalist
4,1801-Jefferson,717,1923,43,1801,Jefferson,Thomas,Democratic-Republican
5,1805-Jefferson,804,2380,45,1805,Jefferson,Thomas,Democratic-Republican
6,1809-Madison,535,1261,21,1809,Madison,James,Democratic-Republican


,Text,Types,Tokens,Sentences,Year,President,FirstName,Party
,<chr>,<int>,<int>,<int>,<int>,<chr>,<chr>,<fct>
55,2005-Bush,772,2312,99,2005,Bush,George W.,Republican
56,2009-Obama,938,2689,112,2009,Obama,Barack,Democratic
57,2013-Obama,814,2317,90,2013,Obama,Barack,Democratic
58,2017-Trump,582,1660,89,2017,Trump,Donald J.,Republican
59,2021-Biden,812,2766,229,2021,Biden,Joseph R.,Democratic
60,2025-Trump,1000,3347,177,2025,Trump,Donald J.,Republican




**Types**  : Número de tokens únicos - ntype()

**Tokens** : Número total de tokens – ntoken()



In [5]:
print(ntype(tail(data_corpus_inaugural)))
print(ntoken(tail(data_corpus_inaugural)))

 2005-Bush 2009-Obama 2013-Obama 2017-Trump 2021-Biden 2025-Trump 
       772        938        814        582        812       1000 
 2005-Bush 2009-Obama 2013-Obama 2017-Trump 2021-Biden 2025-Trump 
      2312       2689       2317       1660       2766       3347 


## 🔖 `docnames(corpus)`

Cuando se crea un corpus en `quanteda`, a cada documento se le asigna automáticamente un nombre o identificador único. Estos identificadores se almacenan en el atributo `docnames` del objeto `corpus`.

---

### 🧰 ¿Para qué sirve?

- Permite **identificar y acceder rápidamente** a documentos individuales dentro del corpus.
- Es útil para aplicar funciones de análisis o visualización **a documentos específicos**.
- Facilita la **trazabilidad** de los resultados, especialmente cuando se trabaja con textos de múltiples fuentes o categorías.

---

📌 Por ejemplo, en el corpus `data_corpus_inaugural`, los `docnames` corresponden al año y presidente de cada discurso inaugural (e.g., `"1789-Washington"`), lo que permite localizar fácilmente un documento específico.


In [6]:
docnames(data_corpus_inaugural)  |> 
  head()  |> 
  print()

[1] "1789-Washington" "1793-Washington" "1797-Adams"      "1801-Jefferson" 
[5] "1805-Jefferson"  "1809-Madison"   


## 🗂️ `meta(corpus)`

La función `meta(corpus)` permite acceder a los **metadatos** asociados al corpus completo. Estos metadatos proporcionan información adicional que describe el corpus en su conjunto (no documentos individuales).

---

### 📌 Características clave:

- Los metadatos se almacenan en forma de una **lista (`list`)**.
- Pueden incluir detalles como:
  - Fuente del corpus
  - Fecha de creación
  - Información de licenciamiento
  - Descripción general del contenido
- Es especialmente útil cuando se trabaja con corpus personalizados o importados desde fuentes externas.

---

🔎 **Ejemplo de uso**:
```r
meta(corpus)


In [7]:
print(meta(data_corpus_inaugural))

$description
[1] "Transcripts of all inaugural addresses delivered by United States Presidents, from Washington 1789 onward.  Data compiled by Gerhard Peters."

$source
[1] "Gerhard Peters and John T. Woolley. The American Presidency Project."

$url
[1] "https://www.presidency.ucsb.edu/documents/presidential-documents-archive-guidebook/inaugural-addresses"

$author
[1] "(various US Presidents)"

$keywords
[1] "political"     "US politics"   "United States" "presidents"   
[5] "presidency"   

$title
[1] "US presidential inaugural address speeches"



---

---


# **II. Tokenization**

## ✂️ ¿Qué es la Tokenización?

La **tokenización** es un proceso fundamental en el procesamiento del lenguaje natural (NLP) que consiste en dividir un texto en unidades más pequeñas llamadas **tokens**.

Un **token** puede ser:

- Una palabra.
- Un número.
- Un signo de puntuación.
- Un fragmento de palabra.
- O cualquier otra unidad significativa del texto.

---

### 🎯 Objetivo

Transformar el texto crudo en una representación estructurada que pueda ser procesada por algoritmos computacionales.

---

### 🧠 ¿Por qué es importante?

- Es el primer paso de prácticamente todos los flujos de trabajo de NLP.
- Permite limpiar y normalizar textos.
- Facilita el cálculo de frecuencias, similitudes y representaciones vectoriales.
- Constituye la base para construir matrices documento-término (DFM), embeddings y modelos de lenguaje.

---

### 🤖 ¿Los LLM utilizan tokenización?

Sí. Los modelos modernos basados en Transformers también tokenizan el texto antes de procesarlo.

La diferencia es que los tokenizadores tradicionales suelen trabajar con palabras completas, mientras que los LLM utilizan frecuentemente **"subpalabras"** (*subword tokens*), lo que les permite manejar vocabulario muy amplio, palabras desconocidas y múltiples idiomas.

Por ejemplo:

```text
Texto:
"La empresa aumentó sus utilidades"

Tokenización tradicional:
["la", "empresa", "aumentó", "sus", "utilidades"]

Tokenización de un LLM:
["La", "empresa", "aument", "ó", "sus", "util", "idades"]
```

---

📌 En **Quanteda**, la tokenización se realiza habitualmente mediante:

```r
tokens(corpus)
```

y constituye el paso previo a la construcción de una matriz documento-término (DFM).

---

## Un Primer Ejemplo






Consideremos el siguiente texto:

"Mario es un gran empresario. Creó su propia empresa a partir de un modesto emprendimiento."

In [8]:
my_text <- c(
"Mario es un gran empresario. Creó su propia empresa a partir de un modesto emprendimiento."
)

Generamos un *Corpus*

In [9]:
my_corpus <- corpus(my_text)
print(my_corpus)
class(my_corpus)

Corpus consisting of 1 document.
text1 :
"Mario es un gran empresario. Creó su propia empresa a partir..."



[1] "corpus"    "character"

Aplicamos la función *toks()* al corpus:

In [10]:
toks <- tokens(my_corpus)
print(toks)
class(toks)

Tokens consisting of 1 document.
text1 :
 [1] "Mario"      "es"         "un"         "gran"       "empresario"
 [6] "."          "Creó"       "su"         "propia"     "empresa"   
[11] "a"          "partir"    
[ ... and 5 more ]



[1] "tokens"

Podemos cambiar todas las palabras a minúsculas - *tokens_tolower()*:

In [11]:
toks_lower <- tokens_tolower(toks)
print(toks_lower)

Tokens consisting of 1 document.
text1 :
 [1] "mario"      "es"         "un"         "gran"       "empresario"
 [6] "."          "creó"       "su"         "propia"     "empresa"   
[11] "a"          "partir"    
[ ... and 5 more ]



Y elminar *stop words* y puntuación:

In [12]:
toks_nopunct_stop <-tokens(toks_lower, remove_punct = TRUE) |>
  tokens_remove(pattern = stopwords("es"))

toks_nopunct_stop 

Tokens consisting of 1 document.
text1 :
[1] "mario"          "gran"           "empresario"     "creó"          
[5] "propia"         "empresa"        "partir"         "modesto"       
[9] "emprendimiento"


Podemos hacer *stemming* (el stemming es una técnica para reducir una palabra a su raíz o “stem”, es decir, a una forma básica que representa todas sus variantes *flexionadas*.)

In [13]:
toks_nopunct_stop_stem <- tokens_wordstem(toks_nopunct_stop)
toks_nopunct_stop_stem

Tokens consisting of 1 document.
text1 :
[1] "mario"          "gran"           "empresario"     "creó"          
[5] "propia"         "empresa"        "partir"         "modesto"       
[9] "emprendimiento"


### ⚠️ El stemming por defecto no funciona en español

El resultado anterior no se ve como una raíz razonable: `tokens_wordstem()` usa por defecto el algoritmo de Porter para **inglés**, por lo que palabras como *"empresario"* o *"emprendimiento"* no se reducen correctamente.

Para aplicar stemming en español (y en muchos otros idiomas), `quanteda` se apoya en el paquete **SnowballC**, que implementa el algoritmo de Snowball para múltiples idiomas.


In [14]:
#install.packages("SnowballC")
library(SnowballC)

In [15]:
toks_nopunct_stop_stem <- tokens_wordstem(toks_nopunct_stop, language = "spanish")
toks_nopunct_stop_stem

Tokens consisting of 1 document.
text1 :
[1] "mari"      "gran"      "empresari" "cre"       "propi"     "empres"   
[7] "part"      "modest"    "emprend"  


---

## 🌿 Stemming vs. Lematización

- **Stemming**: trunca la palabra a una raíz aproximada, que muchas veces no es una palabra real del idioma.
  Ejemplo (en nuestra frase): `empresario` → `empresari`, `empresa` → `empres`, `emprendimiento` → `emprend`.

- **Lematización**: reduce cada palabra a su **forma de diccionario** (lema): el infinitivo de los verbos, el masculino singular de los adjetivos, etc.
  Ejemplo (en nuestra frase): `creó` → `crear` (conjugación → infinitivo), `gran` → `grande` (apócope → forma plena), `propia` → `propio` (femenino → forma de diccionario).

`quanteda` no incluye un lematizador propio, porque lematizar correctamente requiere conocer la **categoría gramatical (POS)** de cada palabra. Para eso existen paquetes especializados:

- **R**: `udpipe`, `spacyr`
- **Python**: `spaCy`, `stanza`

Para ilustrar la diferencia conceptual sin instalar modelos pesados, construimos a mano un pequeño **diccionario de lematización** y lo aplicamos con `tokens_replace()`:


In [16]:
# Diccionario de lematización manual (solo para ilustrar el concepto)
lemma_dict <- c(
  "creó" = "crear",
  "gran" = "grande",
  "propia" = "propio"
)

toks_lemma <- tokens_replace(
  toks_nopunct_stop,
  pattern = names(lemma_dict),
  replacement = lemma_dict,
  valuetype = "fixed"
)

print(toks_nopunct_stop_stem)  # stemming
print(toks_lemma)              # lematización (manual)


Tokens consisting of 1 document.
text1 :
[1] "mari"      "gran"      "empresari" "cre"       "propi"     "empres"   
[7] "part"      "modest"    "emprend"  

Tokens consisting of 1 document.
text1 :
[1] "mario"          "grande"         "empresario"     "crear"         
[5] "propio"         "empresa"        "partir"         "modesto"       
[9] "emprendimiento"



> 🔍 **Comparación**: el stemming reduce `empresario`, `empresa` y `emprendimiento` a raíces como `empresari`, `empres` y `emprend`, que no son palabras del español. La lematización, en cambio, normaliza `creó` → `crear`, `gran` → `grande` y `propia` → `propio`: formas que sí existen en el diccionario, pero que requieren reglas (o un modelo lingüístico) capaces de reconocer conjugaciones verbales, apócopes y concordancia de género/número.
>
> En un corpus real, este diccionario se construiría automáticamente con `udpipe::udpipe_annotate()`, que etiqueta cada token con su lema y su categoría gramatical (POS), en lugar de definirse a mano.


---

### 🤖 ¿Y qué ocurre en los LLM?

Los modelos modernos basados en **Transformers** normalmente no requieren aplicar *stemming* o *lematización* antes del análisis.

Esto se debe a que utilizan mecanismos de tokenización basados en **subpalabras** y representaciones contextuales (*embeddings*), capaces de capturar similitudes entre términos relacionados sin necesidad de reducir explícitamente cada palabra a una raíz común.

Por ejemplo, palabras como:

```text
empresa
empresas
empresarial
empresario
```

suelen ser interpretadas por un LLM como conceptos estrechamente relacionados, incluso sin aplicar stemming.

> 📌 El stemming sigue siendo muy útil en enfoques tradicionales basados en frecuencias, matrices documento-término (DFM) y TF-IDF, donde reducir la variabilidad léxica puede mejorar la calidad del análisis.

---
---

# 🏛️ **III. Discursos Inaugurales**

In [17]:
toks <- tokens(data_corpus_inaugural)
print(toks)
class(toks)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "Fellow-Citizens" "of"              "the"             "Senate"         
 [5] "and"             "of"              "the"             "House"          
 [9] "of"              "Representatives" ":"               "Among"          
[ ... and 1,525 more ]

1793-Washington :
 [1] "Fellow"   "citizens" ","        "I"        "am"       "again"   
 [7] "called"   "upon"     "by"       "the"      "voice"    "of"      
[ ... and 135 more ]

1797-Adams :
 [1] "When"      "it"        "was"       "first"     "perceived" ","        
 [7] "in"        "early"     "times"     ","         "that"      "no"       
[ ... and 2,565 more ]

1801-Jefferson :
 [1] "Friends"   "and"       "Fellow"    "Citizens"  ":"         "Called"   
 [7] "upon"      "to"        "undertake" "the"       "duties"    "of"       
[ ... and 1,911 more ]

1805-Jefferson :
 [1] "Proceeding"    ","             "fellow"        "citizens"     
 [5] ","             "to

[1] "tokens"

---

## ✂️ Tokens, Guiones y Palabras Compuestas

Cuando analizamos texto con `quanteda`, es común encontrar **palabras compuestas con guiones**, como:

> `"Fellow-Citizens"`, `"God-given"`, `"non-believers"`, `"self-reliance"`, `"self-supporting"`, `"well-spent"`, `"well-regulated"`...

---

### 🔧 `tokens()` y el argumento `split_hyphens`

El argumento `split_hyphens` de la función `tokens()` controla si las palabras con guiones deben tratarse como un solo token o dividirse:

- `split_hyphens = FALSE` (por defecto): mantiene las palabras con guiones como **un solo token**.  
  Ejemplo: `"God-given"` → `"God-given"`

- `split_hyphens = TRUE`: divide las palabras compuestas en múltiples **tokens separados**, eliminando el guion.  
  Ejemplo: `"God-given"` → `"God"` + `"given"`

---

### 🎯 ¿Por qué importa?

- Afecta el **conteo de frecuencia** de términos.
- Impacta el resultado de técnicas como **stemming**, **keyness**, o **modelamiento de tópicos**.
- Puede ser útil para capturar **patrones léxicos** de prefijos como `non-`, `self-`, `well-`, etc.

> 🔍 La decisión de dividir o mantener palabras compuestas debe basarse en los **objetivos del análisis**.


---

### 🤖 ¿Qué ocurre en los LLM?

Los modelos modernos basados en **Transformers** también tokenizan las palabras compuestas, pero normalmente no requieren que el investigador decida explícitamente si un término debe dividirse o mantenerse unido.

Por ejemplo:

```text
self-reliance
```

puede ser segmentado internamente en varias subpalabras o símbolos según el tokenizador utilizado por el modelo.

Gracias a los **embeddings contextuales**, los LLM suelen conservar la relación semántica entre los componentes de la palabra compuesta sin necesidad de reglas manuales como `split_hyphens = TRUE`.

> 📌 En NLP tradicional, la decisión de dividir o mantener palabras compuestas puede afectar significativamente las frecuencias, la DFM y los modelos posteriores. En los LLM, esta decisión suele ser manejada automáticamente por el tokenizador del modelo.

---

In [18]:
# Hypehns ----
toks_split <- tokens(data_corpus_inaugural, split_hyphens = TRUE)
print(toks_split)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "Fellow"          "-"               "Citizens"        "of"             
 [5] "the"             "Senate"          "and"             "of"             
 [9] "the"             "House"           "of"              "Representatives"
[ ... and 1,527 more ]

1793-Washington :
 [1] "Fellow"   "citizens" ","        "I"        "am"       "again"   
 [7] "called"   "upon"     "by"       "the"      "voice"    "of"      
[ ... and 135 more ]

1797-Adams :
 [1] "When"      "it"        "was"       "first"     "perceived" ","        
 [7] "in"        "early"     "times"     ","         "that"      "no"       
[ ... and 2,571 more ]

1801-Jefferson :
 [1] "Friends"   "and"       "Fellow"    "Citizens"  ":"         "Called"   
 [7] "upon"      "to"        "undertake" "the"       "duties"    "of"       
[ ... and 1,919 more ]

1805-Jefferson :
 [1] "Proceeding"    ","             "fellow"        "citizens"     
 [5] ","             "to

In [19]:
sum(ntoken(toks))
sum(ntoken(toks_split))

[1] 154888

[1] 155468



---



## 🔠 Mayúsculas, Signos de Puntuación y *Stop Words*

Una vez que hemos tokenizado los textos, es común aplicar transformaciones adicionales para normalizar los datos y reducir la variabilidad léxica.

### 🔡 Conversión a minúsculas

Usamos `tokens_tolower()` para estandarizar las palabras y evitar que términos como `"Government"` y `"government"` se traten como elementos diferentes.

### ❌ Eliminación de puntuación y símbolos

Al tokenizar, podemos utilizar los argumentos `remove_punct = TRUE` y `remove_symbols = TRUE` para eliminar caracteres como comas, puntos, signos y símbolos especiales.

### 🧹 Eliminación de *Stop Words*

Las *stopwords* (como `"the"`, `"and"` o `"in"`) son palabras muy frecuentes que suelen aportar poca información temática. Mediante `tokens_remove()` pueden eliminarse para reducir ruido y simplificar el análisis.

---

### 🎯 ¿Por qué realizar estas transformaciones?

En los enfoques tradicionales basados en frecuencias, matrices documento-término (DFM) y TF-IDF, estas operaciones permiten:

- Reducir la dimensionalidad de los datos.
- Concentrar la información en términos más relevantes.
- Mejorar la interpretación de frecuencias y tópicos.
- Facilitar la comparación entre documentos.

Por ello, constituyen una práctica habitual en muchos flujos de trabajo de NLP clásico.

---



> 📌 En NLP tradicional estas transformaciones suelen mejorar los resultados al reducir la dimensionalidad de los datos. En los LLM, una limpieza excesiva puede eliminar información valiosa para la comprensión del contexto.
---


---

### 🤖 ¿Qué ocurre en los LLM?

En los enfoques tradicionales basados en frecuencias y matrices documento-término (DFM), estas transformaciones suelen mejorar el análisis al reducir la variabilidad léxica y disminuir la dimensionalidad de los datos.

Sin embargo, los modelos modernos basados en **Transformers** normalmente trabajan con el texto casi en su forma original. Las mayúsculas, la puntuación y muchas palabras funcionales pueden contener información relevante para interpretar correctamente el contexto.

Eliminar estos elementos puede:

- **Distorsionar el significado** (por ejemplo, eliminar “no” en “no feliz” cambia completamente el sentido),
- **Romper la sintaxis**, lo que dificulta la comprensión del contexto por parte del modelo,
- **Reducir el rendimiento** en tareas como clasificación, análisis de sentimiento, QA, etc.

Por esta razón, los LLM suelen conservar gran parte de la información original del texto y requieren mucho menos preprocesamiento que los enfoques clásicos.

> 📌 Una regla práctica es que mientras los métodos tradicionales suelen beneficiarse de una limpieza agresiva del texto, los Transformers y LLM generalmente funcionan mejor con textos cercanos a su forma original.
---

In [20]:
toks_lower <- tokens_tolower(toks)
print(toks_lower)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "fellow-citizens" "of"              "the"             "senate"         
 [5] "and"             "of"              "the"             "house"          
 [9] "of"              "representatives" ":"               "among"          
[ ... and 1,525 more ]

1793-Washington :
 [1] "fellow"   "citizens" ","        "i"        "am"       "again"   
 [7] "called"   "upon"     "by"       "the"      "voice"    "of"      
[ ... and 135 more ]

1797-Adams :
 [1] "when"      "it"        "was"       "first"     "perceived" ","        
 [7] "in"        "early"     "times"     ","         "that"      "no"       
[ ... and 2,565 more ]

1801-Jefferson :
 [1] "friends"   "and"       "fellow"    "citizens"  ":"         "called"   
 [7] "upon"      "to"        "undertake" "the"       "duties"    "of"       
[ ... and 1,911 more ]

1805-Jefferson :
 [1] "proceeding"    ","             "fellow"        "citizens"     
 [5] ","             "to



Eliminar los signos de puntuación y las “stop words”



In [21]:
toks_nopunct_stop <-tokens(data_corpus_inaugural, remove_punct = TRUE) |>
  tokens_remove(pattern = stopwords("en"))
print(toks_nopunct_stop)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "Fellow-Citizens" "Senate"          "House"           "Representatives"
 [5] "Among"           "vicissitudes"    "incident"        "life"           
 [9] "event"           "filled"          "greater"         "anxieties"      
[ ... and 640 more ]

1793-Washington :
 [1] "Fellow"     "citizens"   "called"     "upon"       "voice"     
 [6] "country"    "execute"    "functions"  "Chief"      "Magistrate"
[11] "occasion"   "proper"    
[ ... and 50 more ]

1797-Adams :
 [1] "first"       "perceived"   "early"       "times"       "middle"     
 [6] "course"      "America"     "remained"    "unlimited"   "submission" 
[11] "foreign"     "legislature"
[ ... and 1,058 more ]

1801-Jefferson :
 [1] "Friends"   "Fellow"    "Citizens"  "Called"    "upon"      "undertake"
 [7] "duties"    "first"     "executive" "office"    "country"   "avail"    
[ ... and 801 more ]

1805-Jefferson :
 [1] "Proceeding"    "fellow"        "ci

¿Qué palabras considera Quanteda como "stop words"

In [22]:
stopwords_en <- stopwords("en")

In [23]:
print(stopwords_en)

  [1] "i"          "me"         "my"         "myself"     "we"        
  [6] "our"        "ours"       "ourselves"  "you"        "your"      
 [11] "yours"      "yourself"   "yourselves" "he"         "him"       
 [16] "his"        "himself"    "she"        "her"        "hers"      
 [21] "herself"    "it"         "its"        "itself"     "they"      
 [26] "them"       "their"      "theirs"     "themselves" "what"      
 [31] "which"      "who"        "whom"       "this"       "that"      
 [36] "these"      "those"      "am"         "is"         "are"       
 [41] "was"        "were"       "be"         "been"       "being"     
 [46] "have"       "has"        "had"        "having"     "do"        
 [51] "does"       "did"        "doing"      "would"      "should"    
 [56] "could"      "ought"      "i'm"        "you're"     "he's"      
 [61] "she's"      "it's"       "we're"      "they're"    "i've"      
 [66] "you've"     "we've"      "they've"    "i'd"        "you'd"     
 [71] 

In [24]:
stopwords_sp <- stopwords("spanish")
stopwords_sp

[1] "de"           "la"           "que"          "el"           "en"          
  [6] "y"            "a"            "los"          "del"          "se"          
 [11] "las"          "por"          "un"           "para"         "con"         
 [16] "no"           "una"          "su"           "al"           "lo"          
 [21] "como"         "más"          "pero"         "sus"          "le"          
 [26] "ya"           "o"            "este"         "sí"           "porque"      
 [31] "esta"         "entre"        "cuando"       "muy"          "sin"         
 [36] "sobre"        "también"      "me"           "hasta"        "hay"         
 [41] "donde"        "quien"        "desde"        "todo"         "nos"         
 [46] "durante"      "todos"        "uno"          "les"          "ni"          
 [51] "contra"       "otros"        "ese"          "eso"          "ante"        
 [56] "ellos"        "e"            "esto"         "mí"           "antes"       
 [61] "algunos"      "qué"          "unos"         "yo"           "otro"        
 [66] "otras"        "otra"         "él"           "tanto"        "esa"         
 [71] "estos"        "mucho"        "quienes"      "nada"         "muchos"      
 [76] "cual"         "poco"         "ella"         "estar"        "estas"       
 [81] "algunas"      "algo"         "nosotros"     "mi"           "mis"         
 [86] "tú"           "te"           "ti"           "tu"           "tus"         
 [91] "ellas"        "nosotras"     "vosotros"     "vosotras"     "os"          
 [96] "mío"          "mía"          "míos"         "mías"         "tuyo"        
[101] "tuya"         "tuyos"        "tuyas"        "suyo"         "suya"        
[106] "suyos"        "suyas"        "nuestro"      "nuestra"      "nuestros"    
[111] "nuestras"     "vuestro"      "vuestra"      "vuestros"     "vuestras"    
[116] "esos"         "esas"         "estoy"        "estás"        "está"        
[121] "estamos"      "estáis"       "están"        "esté"         "estés"       
[126] "estemos"      "estéis"       "estén"        "estaré"       "estarás"     
[131] "estará"       "estaremos"    "estaréis"     "estarán"      "estaría"     
[136] "estarías"     "estaríamos"   "estaríais"    "estarían"     "estaba"      
[141] "estabas"      "estábamos"    "estabais"     "estaban"      "estuve"      
[146] "estuviste"    "estuvo"       "estuvimos"    "estuvisteis"  "estuvieron"  
[151] "estuviera"    "estuvieras"   "estuviéramos" "estuvierais"  "estuvieran"  
[156] "estuviese"    "estuvieses"   "estuviésemos" "estuvieseis"  "estuviesen"  
[161] "estando"      "estado"       "estada"       "estados"      "estadas"     
[166] "estad"        "he"           "has"          "ha"           "hemos"       
[171] "habéis"       "han"          "haya"         "hayas"        "hayamos"     
[176] "hayáis"       "hayan"        "habré"        "habrás"       "habrá"       
[181] "habremos"     "habréis"      "habrán"       "habría"       "habrías"     
[186] "habríamos"    "habríais"     "habrían"      "había"        "habías"      
[191] "habíamos"     "habíais"      "habían"       "hube"         "hubiste"     
[196] "hubo"         "hubimos"      "hubisteis"    "hubieron"     "hubiera"     
[201] "hubieras"     "hubiéramos"   "hubierais"    "hubieran"     "hubiese"     
[206] "hubieses"     "hubiésemos"   "hubieseis"    "hubiesen"     "habiendo"    
[211] "habido"       "habida"       "habidos"      "habidas"      "soy"         
[216] "eres"         "es"           "somos"        "sois"         "son"         
[221] "sea"          "seas"         "seamos"       "seáis"        "sean"        
[226] "seré"         "serás"        "será"         "seremos"      "seréis"      
[231] "serán"        "sería"        "serías"       "seríamos"     "seríais"     
[236] "serían"       "era"          "eras"         "éramos"       "erais"       
[241] "eran"         "fui"          "fuiste"       "fue"          "fuimos"      
[246] "fuisteis"     "fueron"    

---

### 🚫 Limpiezas tradicionales que **deben evitarse** con LLMs

| Paso                       | ¿Por qué evitarlo?                                                  |
|----------------------------|---------------------------------------------------------------------|
| Eliminación de stopwords   | Afecta la estructura y el significado de las oraciones.             |
| Eliminación de puntuación  | Rompe los límites y el ritmo de las frases.                         |
| Stemming/Lematización      | Puede generar palabras no naturales o no reconocidas por el modelo. |
| Convertir a minúsculas     | Algunos modelos son sensibles a mayúsculas/minúsculas (cased).      |

---

### ✅ ¿Cuándo sí aplicar una limpieza mínima?

En algunos casos puede ser útil:

- Eliminar etiquetas HTML, URLs o caracteres corruptos.
- Normalizar espacios en blanco o codificación.
- Quitar ruido no lingüístico.

Pero en general, se debe **preservar la estructura lingüística natural** del texto.

---

### 📌 Excepción:
Si se usan métodos tradicionales como **TF-IDF**, **modelos de tópicos (LDA)** o **clasificadores clásicos**, entonces **sí se recomienda una limpieza más intensa**.



### ¿Por qué eliminar las "stop words"?



1.   Reducción del ruido: Las stopwords son palabras muy comunes en un idioma determinado, como "el", "de", "y", "a", etc. Estas palabras suelen tener poca información semántica y, en muchos casos, no aportan un valor significativo al análisis de texto. Al eliminar las stopwords, se reduce el ruido en los datos y se enfoca en las palabras más relevantes para el análisis.

2.   Mejora de la eficiencia computacional: Las stopwords son palabras que aparecen con mucha frecuencia en los documentos. Al eliminarlas, se reduce la cantidad de palabras en el texto, lo que a su vez reduce la cantidad de cálculos y operaciones necesarias durante el análisis. Esto puede acelerar el procesamiento y mejorar la eficiencia computacional.


3.   Enfocarse en el contexto y las palabras clave: Al eliminar las stopwords, se puede poner mayor énfasis en las palabras que tienen más peso semántico y contribuyen significativamente al significado y al contexto de un texto. Esto puede ayudar a identificar y analizar las palabras clave, los patrones y las relaciones más relevantes en el análisis de texto.

Es importante destacar que la eliminación de stopwords no es aplicable en todos los casos y en todos los tipos de análisis de texto. En algunos casos, como el análisis de contexto o la identificación de temas generales, las stopwords pueden contener información relevante y no deben eliminarse. Por lo tanto, es esencial considerar el contexto y los objetivos del análisis antes de decidir si es apropiado eliminar o conservar las stopwords.




In [25]:
toks_nopunct_stop_low <- tokens(data_corpus_inaugural, remove_punct = TRUE,)  |>
  tokens_remove(pattern = stopwords("en"))  |>
  tokens_tolower()

print(toks_nopunct_stop_low)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "fellow-citizens" "senate"          "house"           "representatives"
 [5] "among"           "vicissitudes"    "incident"        "life"           
 [9] "event"           "filled"          "greater"         "anxieties"      
[ ... and 640 more ]

1793-Washington :
 [1] "fellow"     "citizens"   "called"     "upon"       "voice"     
 [6] "country"    "execute"    "functions"  "chief"      "magistrate"
[11] "occasion"   "proper"    
[ ... and 50 more ]

1797-Adams :
 [1] "first"       "perceived"   "early"       "times"       "middle"     
 [6] "course"      "america"     "remained"    "unlimited"   "submission" 
[11] "foreign"     "legislature"
[ ... and 1,058 more ]

1801-Jefferson :
 [1] "friends"   "fellow"    "citizens"  "called"    "upon"      "undertake"
 [7] "duties"    "first"     "executive" "office"    "country"   "avail"    
[ ... and 801 more ]

1805-Jefferson :
 [1] "proceeding"    "fellow"        "ci




---


## 🌱 IV. Stemming

El **stemming** es el proceso de reducir las palabras a su forma básica o raíz léxica, llamada *stem*. Esta raíz no siempre es una palabra válida, pero permite agrupar variantes morfológicas como si fueran equivalentes. Por ejemplo:

- *run*, *runner*, *running* → **run**
- *connect*, *connected*, *connecting*, *connection* → **connect**

Este proceso es útil para:

- Reducir la dimensión del vocabulario.
- Identificar similitudes semánticas entre términos relacionados.
- Mejorar tareas como el análisis de frecuencia, modelamiento de tópicos o recuperación de información.

En `quanteda`, el stemming se aplica con la función `tokens_wordstem()`. Es especialmente útil cuando queremos enfocarnos en el contenido general de un texto, sin preocuparnos por sus variaciones gramaticales.

> ⚠️ A diferencia de la **lematización**, el stemming no devuelve necesariamente palabras del diccionario, sino formas truncadas útiles para análisis computacional.


---


In [26]:
tokens_wordstem(toks_nopunct_stop_low)

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "fellow-citizen" "senat"          "hous"           "repres"        
 [5] "among"          "vicissitud"     "incid"          "life"          
 [9] "event"          "fill"           "greater"        "anxieti"       
[ ... and 640 more ]

1793-Washington :
 [1] "fellow"   "citizen"  "call"     "upon"     "voic"     "countri" 
 [7] "execut"   "function" "chief"    "magistr"  "occas"    "proper"  
[ ... and 50 more ]

1797-Adams :
 [1] "first"      "perceiv"    "earli"      "time"       "middl"     
 [6] "cours"      "america"    "remain"     "unlimit"    "submiss"   
[11] "foreign"    "legislatur"
[ ... and 1,058 more ]

1801-Jefferson :
 [1] "friend"   "fellow"   "citizen"  "call"     "upon"     "undertak"
 [7] "duti"     "first"    "execut"   "offic"    "countri"  "avail"   
[ ... and 801 more ]

1805-Jefferson :
 [1] "proceed"   "fellow"    "citizen"   "qualif"    "constitut" "requir"   
 [7] "entranc"   "charg"    

En cuanto a las limitaciones inherentes al stemming,  probablemente el problema más importante está relacionado con el **over-stemming** (situación en la que dos palabras no relacionadas se reducen a la misma raíz) y el **under-stemming** (situación en la que dos palabras relacionadas se reducen a diferentes raíces). En este sentido, trabajar con la versión *stemmed* de un corpus requiere más intervención "humana" para que el analista pueda evaluar correctamente el significado correcto de una palabra **stemmed**.
(*over-stemming* y  *under-stemming* me recuerdan el trade-off entre *over-fitting* y *under-fitting* en modelos de ML).



---
# V. Key Words in Context - kwic()

### 📌 ¿Qué hace `kwic()`?

`kwic()` es una técnica clásica de análisis de texto que muestra una palabra clave rodeada por un número definido de palabras antes y después. Es útil para observar cómo se usa un término en distintos contextos lingüísticos.

In [27]:
toks <- tokens(data_corpus_inaugural)
kw_inaug <- kwic(toks, pattern = c("liberty", "people"), window = 4)
print(tail(kw_inaug, 10))



Keyword-in-context with 10 matches.                                                        
 [2021-Biden, 2142]     them by becoming the | people  |
 [2021-Biden, 2681] That our America secured | liberty |
  [2025-Trump, 440]     importantly, its own | people  |
  [2025-Trump, 461]   shown by the wonderful | people  |
  [2025-Trump, 692]          and to give the | people  |
 [2025-Trump, 1058]       won by millions of | people  |
 [2025-Trump, 2808]     cannot achieve. Many | people  |
 [2025-Trump, 2835]         am. The American | people  |
 [2025-Trump, 3047]             . We are one | people  |
 [2025-Trump, 3201]      again, including by | people  |
                          
 and nation we know       
 at home and stood        
 . Our country can        
 of North Carolina —      
 back their faith,        
 . To the Black           
 thought it was impossible
 have spoken. I           
 , one family,            
 of religion, faith       



Nótese que en este caso, se utiliza la versión más básica del texto tokenizado.  Al utilizar la versión básica del texto tokenizado, se pueden presentar las palabras clave junto con su contexto de manera más clara en un lenguage que hace más sencilla la interpretación humana.

Para exportar el resultado:

In [28]:
library(writexl)

In [29]:
kw_df <- as.data.frame(kw_inaug)
kw_df

docname,from,to,pre,keyword,post,pattern
<chr>,<int>,<int>,<chr>,<chr>,<chr>,<fct>
1789-Washington,430,430,and happiness of the,people,of the United States,people
1789-Washington,507,507,than either . No,people,can be bound to,people
1789-Washington,1037,1037,the sacred fire of,liberty,and the destiny of,liberty
1789-Washington,1070,1070,hands of the American,people,. Besides the ordinary,people
1789-Washington,1473,1473,to favor the American,people,with opportunities for deliberating,people
1793-Washington,60,60,in me by the,people,of united America .,people
1797-Adams,108,108,and intelligence of the,people,", under an overruling",people
1797-Adams,187,187,and ardor of the,people,during the Revolutionary war,people
1797-Adams,255,255,only ones which the,people,at large had ever,people


In [ ]:

# Exportar a archivo Excel
write_xlsx(kw_df, path = "kwic_output.xlsx")

### 🔍 ¿Pueden los LLMs realizar tareas tipo `kwic()`?

### ✅ Sí, pero no es su principal fortaleza.

Los **modelos de lenguaje grande (LLMs)** pueden identificar ocurrencias de palabras o expresiones dentro de un texto y mostrar el contexto que las rodea, realizando una tarea conceptualmente similar a `kwic()` (*Key Word In Context*).

Sin embargo, los LLMs fueron diseñados para comprender e interpretar lenguaje, no para realizar búsquedas estructuradas sobre corpus extensos.

---

### 🤖 ¿Qué puede hacer un LLM?

Un LLM puede:

- Identificar una palabra o expresión en un texto.
- Resumir el contexto en el que aparece.
- Explicar diferencias de significado según el contexto.
- Agrupar usos similares de un término.

Por ejemplo, un LLM puede responder preguntas como:

> "¿Cómo se utiliza la palabra *government* a lo largo de estos discursos?"

o

> "¿En qué contextos aparece la palabra *inflation* en este conjunto de documentos?"

---

### ⚠️ ¿Qué limitaciones tiene?

Para tareas de búsqueda sistemática sobre corpus grandes, los LLM presentan algunas limitaciones:

- Pueden omitir ocurrencias.
- No garantizan reproducibilidad perfecta.
- Son menos eficientes para recorrer miles de documentos.
- Resultan innecesariamente costosos para tareas puramente mecánicas de búsqueda.

---

### 📊 Comparación: LLM vs `kwic()`

| Aspecto | LLM | `kwic()` |
|----------|------|---------|
| Búsqueda exacta de ocurrencias | ⚠️ Buena, pero no garantizada | ✅ Exacta |
| Escalabilidad | ⚠️ Limitada por contexto y costo | ✅ Muy alta |
| Reproducibilidad | ⚠️ Moderada | ✅ Total |
| Interpretación semántica | ✅ Muy alta | ❌ Limitada |
| Velocidad en corpus grandes | ❌ Menor | ✅ Muy alta |
| Transparencia metodológica | ⚠️ Moderada | ✅ Alta |

---

### 🎯 ¿Cuándo usar cada uno?

Utiliza **`kwic()`** cuando necesites:

- Recuperar todas las ocurrencias de un término.
- Realizar análisis reproducibles.
- Trabajar con corpus grandes.
- Obtener resultados exactos y verificables.

Utiliza un **LLM** cuando necesites:

- Interpretar los contextos encontrados.
- Resumir patrones de uso.
- Comparar significados entre documentos.
- Explorar relaciones semánticas complejas.

> 📌 En la práctica, ambos enfoques son complementarios: `kwic()` permite recuperar evidencia textual de manera sistemática, mientras que los LLMs pueden ayudar a interpretar y sintetizar esa evidencia.

---

# VI. Ejercicio de Aplicación

En este ejercicio aplicarás el flujo completo de limpieza y transformación de textos visto en esta unidad, sobre un subconjunto distinto de `data_corpus_inaugural`.

**Pasos sugeridos:**

1. Selecciona los **últimos 10 discursos** del corpus: `corpus_subset <- tail(data_corpus_inaugural, 10)`.
2. Tokeniza el corpus y calcula `ntoken()` y `ntype()` sobre la versión **sin procesar**.
3. Aplica, en orden: minúsculas → eliminación de puntuación y símbolos → eliminación de stopwords → stemming.
4. Vuelve a calcular `ntoken()` y `ntype()` sobre la versión procesada. ¿En qué porcentaje se redujo el vocabulario (*types*)?
5. Usa `kwic()` para buscar una palabra de tu elección (ej. `"freedom"`, `"war"`, `"economy"`) en la versión sin procesar y en la procesada. ¿Cambia el resultado o su interpretabilidad?
6. **Discusión**: ¿en qué casos esta reducción de vocabulario podría hacer perder información relevante para un análisis posterior con un modelo LLM (ver Unidad 6)?

> 💡 Tip: reutiliza el código de las secciones anteriores como plantilla. El objetivo es practicar el pipeline completo `corpus → tokens → limpieza` y reflexionar sobre sus efectos.


In [ ]:
# Tu código aquí
